embedded_rtic/examples/sonars.rs

```diff
@@ -0,0 +1,448 @@
+            gpiote.channel1().clear();
+            gpiote.channel2().clear();
+        });
+    }
+
+    #[task(priority = 2, local = [trig], shared = [gpiote])]
+    /// Send a small pulse to the sonars.
+    ///
+    /// The sonars will then notify us in echo when the sound waves are
+    /// recieved.
+    async fn trigger1(cx: trigger1::Context) {
+        // Set high is always valid.
+        cx.local.trig.set_high().unwrap();
+
+        let now = Mono::now();
+        Mono::delay_until(now + 20.micros()).await;
+
+        // Set low is always valid.
+        cx.local.trig.set_low().unwrap();
+    }
+
+    #[task(priority = 2, local = [trig2], shared = [gpiote])]
+    /// Send a small pulse to the sonars.
+    ///
+    /// The sonars will then notify us in echo when the sound waves are
+    /// recieved.
+    async fn trigger2(cx: trigger2::Context) {
+        // Set high is always valid.
+        cx.local.trig2.set_high().unwrap();
+
+        let now = Mono::now();
+        Mono::delay_until(now + 20.micros()).await;
+
+        // Set low is always valid.
+        cx.local.trig2.set_low().unwrap();
+    }
+
+    #[task(priority = 2, local = [trig3], shared = [gpiote])]
+    /// Send a small pulse to sonar 3.
+    ///
+    /// The sonar will then notify us in echo when the sound wave is recieved.
+    async fn trigger3(cx: trigger3::Context) {
+        // Set high is always valid.
+        cx.local.trig3.set_high().unwrap();
+
+        let now = Mono::now();
+        Mono::delay_until(now + 20.micros()).await;
+
+        // Set low is always valid.
+        cx.local.trig3.set_low().unwrap();
+    }
+
+    #[task(priority = 1,local = [receiver_forward])]
+    /// Re-spawn trigger after every new distance is correctly measured.
+    async fn trigger_trampoline1(cx: trigger_trampoline1::Context) {
+        // Apply a sliding window-like smoothing to the measurements
+        let mut window: ArrayDeque<u32, 20, Wrapping> = ArrayDeque::new();
+        let outlier_offset = 100;
+        let mut prev_avg = u32::MAX - outlier_offset;
+        loop {
+            // Receive data from the first sonar
+            match trigger1::spawn() {
+                Ok(_) => {
+                    let distance = cx.local.receiver_forward.recv().await.unwrap();
+                    // Discard outliers
+                    if distance < prev_avg + outlier_offset {
+                        let _ = window.push_back(distance);
+                    }
+                    let avg = window.iter().sum::<u32>() / (window.len() as u32);
+                    prev_avg = avg;
+
+                    info!(
+                        "Sonar 1 Distance : {:?} cm (average : {:?} cm)",
+                        distance, avg
+                    );
+                }
+                Err(_) => {
+                    error!("Failed to receive data from sonar 1");
+                }
+            }
+            Mono::delay(1000.millis()).await; // Adjust delay as needed
```

Comment:
Would be interesting to see how fast everything goes without these, about how fast do we get the measurements?